core/src/avm2/object/xml_list_object.rs

```diff
@@ -539,39 +539,329 @@ impl<'gc> TObject<'gc> for XmlListObject<'gc> {
-                            target
-                        )
-                        .into());
-                    }
+                    let r = if let Some(target) = self.target_object() {
+                        // 2.a.i. Let r be the result of calling the [[ResolveValue]] method of x.[[TargetObject]]
+                        let r = target.resolve_value(activation)?;
 
-                    if index >= write.children.len() {
-                        if let Some(value_xml) =
-                            value.as_object().and_then(|obj| obj.as_xml_object())
-                        {
-                            write.children.push(E4XOrXml::Xml(value_xml));
+                        // 2.a.ii. If r == null, return
+                        let Some(r) = r else {
                             return Ok(());
+                        };
+
+                        Some(r)
+                    // 2.b. Else let r = null
+                    } else {
+                        None
+                    };
+
+                    // 2.c. If i is greater than or equal to x.[[Length]]
+                    if index >= self.length() {
+                        let r = match r {
+                            Some(XmlOrXmlListObject::Xml(x)) => Some(*x.node()),
+                            // 2.c.i. If Type(r) is XMLList
+                            Some(XmlOrXmlListObject::XmlList(x)) => {
+                                // 2.c.i.1. If r.[[Length]] is not equal to 1, return
+                                if x.length() != 1 {
+                                    return Ok(());
+                                }
+
+                                // 2.c.i.2. Else let r = r[0]
+                                Some(*x.children()[0].node())
+                            }
+                            None => None,
+                        };
+
+                        // 2.c.ii. If r.[[Class]] is not equal to "element", return
+                        if let Some(r) = r {
+                            if !matches!(*r.kind(), E4XNodeKind::Element { .. }) {
+                                return Ok(());
+                            }
+                        }
+
+                        // 2.c.iii Create a new XML object y with y.[[Parent]] = r, y.[[Name]] = x.[[TargetProperty]],
+                        //         y.[[Attributes]] = {}, y.[[Length]] = 0
+                        let y = match self.target_property() {
+                            // 2.c.iv. If Type(x.[[TargetProperty]]) is AttributeName
+                            Some(x) if x.is_attribute() => {
+                                // 2.c.iv.1. Let attributeExists be the result of calling the [[Get]] method of r with argument y.[[Name]]
+                                // FIXME: Looking at avmplus, it seems it does not check if r is null, not sure if we should use a unwrap or if let?
+                                let attribute_exists = XmlObject::new(r.unwrap(), activation)
```

Comment:
Should this be an `expect` at the very least so we can have a little more detail about it in the error report?

Reply:
Yeah, you are right, I don't think there is a case where target_property would be Some and target_object would be None. (It is usually the other way around).